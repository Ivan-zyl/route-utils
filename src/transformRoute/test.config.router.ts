const list = [
  {
    path: '/home',
    component: './Home',
    menuKey: 'home',
    parentKey: '',
  },
  {
    path: '/manageSite',
    component: './platformConfig/ManageSite',
    permission: 'KM_PLAT_ADMIN',
    menuKey: 'manageSite',
    parentKey: 'platformConfig',
  },
  {
    path: '/manageSite/addSite',
    component: './platformConfig/AddSite',
    permission: 'KM_PLAT_ADMIN',
    menuKey: 'manageSite',
    parentKey: 'platformConfig',
  },
  {
    path: '/manageSite/updateSite/:id',
    component: './platformConfig/UpdateSite',
    permission: 'KM_PLAT_ADMIN',
    menuKey: 'manageSite',
    parentKey: 'platformConfig',
  },
  {
    path: '/manageSite/viewSite/:id',
    component: './platformConfig/UpdateSite',
    permission: 'KM_PLAT_ADMIN',
    menuKey: 'manageSite',
    parentKey: 'platformConfig',
  },
  /**
   * 金融云站点租户
   */
  {
    path: '/siteTenant',
    component: './platformConfig/SiteTenant',
    permission: 'KM_PLAT_ADMIN',
    menuKey: 'siteTenant',
    parentKey: 'platformConfig',
  },
  {
    path: '/siteTenant/addTenant',
    component: './platformConfig/AddTenant',
    permission: 'KM_PLAT_ADMIN',
    menuKey: 'siteTenant',
    parentKey: 'platformConfig',
  },
  {
    path: '/siteTenant/updateTenant/:id',
    component: './platformConfig/AddTenant',
    permission: 'KM_PLAT_ADMIN',
    menuKey: 'siteTenant',
    parentKey: 'platformConfig',
  },
  {
    path: '/siteTenant/viewTenant/:id',
    component: './platformConfig/AddTenant',
    permission: 'KM_PLAT_ADMIN',
    menuKey: 'siteTenant',
    parentKey: 'platformConfig',
  },

  /**
   * 安全工程师信息
   */
  {
    path: '/engineer',
    component: './platformConfig/EngineerInfo',
    permission: 'KM_PLAT_ADMIN',
    menuKey: 'engineer',
    parentKey: 'platformConfig',
  },
  /**
   * 密码技术团队人员信息
   */
  {
    path: '/team',
    component: './platformConfig/TeamMembers',
    permission: 'KM_PLAT_ADMIN',
    menuKey: 'team',
    parentKey: 'platformConfig',
  },
  /**
   * 业务安全目标
   */

  {
    path: '/securityGoals',
    component: './platformConfig/SecurityGoals',
    permission: 'KM_PLAT_ADMIN',
    menuKey: 'securityGoals',
    parentKey: 'platformConfig',
  },
  {
    path: '/securityGoals/add',
    component: './platformConfig/AddSecurityGoals',
    permission: 'KM_PLAT_ADMIN',
    menuKey: 'securityGoals',
    parentKey: 'platformConfig',
  },
  {
    path: '/securityGoals/update/:id',
    component: './platformConfig/AddSecurityGoals',
    permission: 'KM_PLAT_ADMIN',
    menuKey: 'securityGoals',
    parentKey: 'platformConfig',
  },

  /**
   * 密钥算法
   */

  {
    path: '/algorithm',
    component: './platformConfig/Algorithm',
    permission: 'KM_PLAT_ADMIN',
    menuKey: 'algorithm',
    parentKey: 'platformConfig',
  },
  {
    path: '/algorithm/add',
    component: './platformConfig/AddAlgorithm',
    permission: 'KM_PLAT_ADMIN',
    menuKey: 'algorithm',
    parentKey: 'platformConfig',
  },
  {
    path: '/algorithm/update/:id',
    component: './platformConfig/AddAlgorithm',
    permission: 'KM_PLAT_ADMIN',
    menuKey: 'algorithm',
    parentKey: 'platformConfig',
  },
  /**
   * 平台配置
   */
  {
    path: '/createKey',
    component: './keyMangement/CreateKey',
    permission: 'KM_WORKFLOW',
    menuKey: 'createKey',
    parentKey: 'keyMangement',
  },
  {
    path: '/useTheExistingKey',
    component: './keyMangement/CreateKey',
    permission: 'KM_WORKFLOW',
    menuKey: 'useTheExistingKey',
    parentKey: 'keyMangement',
  },
  {
    path: '/queryPublicKeyContent',
    component: './keyMangement/QueryKeyContent',
    permission: 'KM_WORKFLOW',
    menuKey: 'queryPublicKeyContent',
    parentKey: 'keyMangement',
  },
  {
    path: '/queryKeyContent',
    component: './keyMangement/QueryKeyContent',
    permission: 'KM_WORKFLOW',
    menuKey: 'queryKeyContent',
    parentKey: 'keyMangement',
  },
  {
    path: '/queryKeyAttr',
    component: './keyMangement/QueryKeyAttr',
    permission: 'KM_WORKFLOW',
    menuKey: 'queryKeyAttr',
    parentKey: 'keyMangement',
  },
  {
    path: '/updateKeyAttr',
    component: './keyMangement/QueryKeyAttr',
    permission: 'KM_WORKFLOW',
    menuKey: 'updateKeyAttr',
    parentKey: 'keyMangement',
  },
  /**
   * 密码服务场景接入
   */
  {
    path: '/sceneAccess',
    component: './scene/SceneAccess',
    permission: 'KM_WORKFLOW',
    menuKey: 'sceneAccess',
    parentKey: 'scene',
  },

  /**
   * 小工具
   */
  {
    path: '/kmiTool',
    component: './keyMangement/KmiTool',
    permission: 'KM_WORKFLOW',
    menuKey: 'kmiTool',
    parentKey: 'tool',
  },
  {
    path: '/smallTool',
    component: './keyMangement/SmallTool',
    permission: 'KM_WORKFLOW',
    menuKey: 'smallTool',
    parentKey: 'tool',
  },
  /**
   * 我的需求单
   */
  {
    path: '/myDemandList',
    component: './demand/MyDemandList',
    permission: 'KM_WORKFLOW',
    menuKey: 'myDemandList',
    parentKey: 'demand',
  },
  {
    path: '/myDemandList/demandDetail/:id',
    component: './demand/DemandDetail',
    permission: 'KM_WORKFLOW',
    menuKey: 'myDemandList',
    parentKey: 'demand',
  },
  {
    path: '/temporaryDemand',
    component: './demand/TemporaryDemand',
    permission: 'KM_WORKFLOW',
    menuKey: 'temporaryDemand',
    parentKey: 'demand',
  },
  {
    path: '/temporaryDemand/detail/:id',
    component: './demand/TemporaryDemandDetail',
    permission: 'KM_WORKFLOW',
    menuKey: 'temporaryDemand',
    parentKey: 'demand',
  },
  /**
   * 我的审批
   */
  {
    path: '/myApproval',
    component: './approval/MyApproval',
    permission: 'KM_WORKFLOW',
    menuKey: 'myApproval',
    parentKey: 'approval',
  },
  {
    path: '/myApproval/approvalDetail/:id',
    component: './approval/ApprovalDetail',
    permission: 'KM_WORKFLOW',
    menuKey: 'myApproval',
    parentKey: 'approval',
  },
  {
    path: '/haveBeenApproval',
    component: './approval/MyApproval',
    permission: 'KM_WORKFLOW',
    menuKey: 'haveBeenApproval',
    parentKey: 'approval',
  },
  {
    path: '/haveBeenApproval/detail/:id',
    component: './approval/ApprovalDetail',
    permission: 'KM_WORKFLOW',
    menuKey: 'haveBeenApproval',
    parentKey: 'approval',
  },
  {
    path: '/claimed',
    component: './approval/DemandClaimed',
    permission: 'KM_WORKFLOW',
    menuKey: 'claimed',
    parentKey: 'approval',
  },
  {
    path: '/claimed/detail/:id',
    component: './approval/ApprovalDetail',
    permission: 'KM_WORKFLOW',
    menuKey: 'claimed',
    parentKey: 'approval',
  },
  {
    path: '/keyManageOfManager',
    component: './keyManageOfManager/KeyManageOfManager',
    permission: 'KM_ADMIN',
    menuKey: 'keyManageOfManager',
  },
  {
    path: '/keyManageOfManager/detail/:id',
    component: './demand/DemandDetail',
    menuKey: 'keyManageOfManager',
    parentKey: '',
  },
  /**
   * secretmng
   */
  {
    path: '/secretmng/commission/secretList',
    component: './secretmng/secretList',
    permission: 'KM_SECRET_ADMIN',
    menuKey: 'secretList',
    parentKey: 'subMenu_manage',
  },
  {
    path: '/secretmng/commission/manualAddSecret',
    component: './secretmng/manualAddSecret',
    permission: 'KM_SECRET_ADMIN',
    menuKey: 'manualAddSecret',
    parentKey: 'subMenu_manage',
  },
  {
    path: '/secretmng/commission/createTask',
    component: './secretmng/createTask',
    permission: 'KM_SECRET_ADMIN',
    menuKey: 'createTask',
    parentKey: 'subMenu_manage',
  },
  {
    path: '/secretmng/commission/componentBls',
    component: './secretmng/componentBls',
    permission: 'KM_SECRET_ADMIN',
    menuKey: 'componentBls',
    parentKey: 'subMenu_manage',
  },
  {
    path: '/secretmng/rotation/taskList',
    component: './secretmng/taskList',
    permission: 'KM_SECRET_ADMIN',
    menuKey: 'taskList',
    parentKey: 'subMenu_rotate',
  },
  {
    path: '/secretmng/rotation/alterPloyManagement',
    component: './secretmng/alterPloyManagement',
    permission: 'KM_SECRET_ADMIN',
    menuKey: 'alterPloyManagement',
    parentKey: 'subMenu_rotate',
  },
  {
    path: '/secretmng/rotation/alterPloyManagement/createAlterPloy',
    component: './secretmng/alterPloyManagement/components/createAlterPloy',
    permission: 'KM_SECRET_ADMIN',
    menuKey: 'createAlterPloy',
    parentKey: 'subMenu_rotate',
  },
  {
    path: '/secretmng/rotation/alterPloyManagement/editAlterPloy',
    component: './secretmng/alterPloyManagement/components/createAlterPloy',
    permission: 'KM_SECRET_ADMIN',
    menuKey: 'editAlterPloy',
    parentKey: 'subMenu_rotate',
  },
  {
    path: '/secretmng/rotation/mailPloyManagement',
    component: './secretmng/mailPloyManagement',
    permission: 'KM_SECRET_ADMIN',
    menuKey: 'mailPloyManagement',
    parentKey: 'subMenu_rotate',
  },
  {
    path: '/secretmng/rotation/mailPloyManagement/createMailPloy',
    component: './secretmng/mailPloyManagement/components/createMailPloy',
    permission: 'KM_SECRET_ADMIN',
    menuKey: 'createMailPloy',
    parentKey: 'subMenu_rotate',
  },
  {
    path: '/secretmng/rotation/mailPloyManagement/editMailPloy',
    component: './secretmng/mailPloyManagement/components/createMailPloy',
    permission: 'KM_SECRET_ADMIN',
    menuKey: 'editMailPloy',
    parentKey: 'subMenu_rotate',
  },
  {
    path: '/secretmng/rotation/blackList',
    component: './secretmng/blackList',
    permission: 'KM_SECRET_ADMIN',
    menuKey: 'blackList',
    parentKey: 'subMenu_rotate',
  },
];

export const routeList = list;
export default {
  path: '/',
  component: '../layout',
  indexRoute: { component: './Home' },
  childRoutes: list,
};
