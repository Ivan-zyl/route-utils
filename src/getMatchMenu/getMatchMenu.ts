import { pathToRegexp } from '@qixian.cs/path-to-regexp';
import { MenuDataItem } from '../types';
import getFlatMenu from '../getFlatMenus/getFlatMenus';
import { isUrl } from '../transformRoute/transformRoute';

/**
 * a-b-c
 * [
 *  "a",
 *  "a-b",
 *  "a-b-c"
 * ]
 * @param menuKey
 */
export const genKeysToArray = (menuKey: string) => {
  const keys = menuKey.split('-');
  const keyArray: string[] = [];
  keys.forEach((key, index) => {
    if (index === 0) {
      keyArray.push(key);
      return;
    }
    keyArray.push(keys.slice(0, index + 1).join('-'));
  });
  return keyArray;
};

export const getMenuMatches = (
  flatMenuKeys: string[] = [],
  path: string,
  flatMenus: {
    [key: string]: MenuDataItem;
  },
): string | undefined =>
  flatMenuKeys
    .filter((item) => {
      // 如果配置了 hideInMenu 就不需要选中了
      if (flatMenus[item].hideInMenu) {
        return false;
      }
      if (item === '/' && path === '/') {
        return true;
      }
      if (item !== '/' && item !== '/*' && item && !isUrl(item)) {
        try {
          // /a
          if (pathToRegexp(`${item}`, []).test(path)) {
            return true;
          }
          // /a/b/b
          if (pathToRegexp(`${item}(.*)`).test(path)) {
            return true;
          }
        } catch (error) {
          console.log(error, path);
        }
      }
      return false;
    })
    .sort((a, b) => {
      // 如果完全匹配放到最后面
      if (a === path) {
        return 10;
      }
      if (b === path) {
        return -10;
      }
      return a.substr(1).split('/').length - b.substr(1).split('/').length;
    })
    .pop();

/**
 * 获取当前的选中菜单列表
 * @param pathname
 * @param menuData
 * @returns MenuDataItem[]
 */
export const getMatchMenu = (
  pathname: string,
  menuData: MenuDataItem[],
): MenuDataItem[] => {
  const flatMenus = getFlatMenu(menuData);
  const flatMenuKeys = Object.keys(flatMenus);
  const menuPathKey = getMenuMatches(flatMenuKeys, pathname || '/', flatMenus);
  if (!menuPathKey) {
    return [];
  }
  const menuItem = flatMenus[menuPathKey] || {
    pro_layout_parentKeys: '',
    key: '',
  };

  // 去重
  const map = new Map();
  const parentItems = (menuItem.pro_layout_parentKeys || [])
    .map((key) => {
      if (map.has(key)) {
        return null;
      }
      map.set(key, true);
      return flatMenus[key];
    })
    .filter((item) => item) as MenuDataItem[];
  if (menuItem.key) {
    parentItems.push(menuItem);
  }
  return parentItems;
};

export default getMatchMenu;
