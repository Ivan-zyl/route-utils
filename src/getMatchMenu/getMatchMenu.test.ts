import getMatchMenu from './getMatchMenu';
import transformRoute from '../transformRoute/transformRoute';
import testMenuData from './test.menu';
import test_config_router2 from './test.config.router2';

const routes = [
  {
    path: '/welcome',
    name: 'welcome',
    exact: true,
    unaccessible: false,
  },
  {
    path: '/admin',
    name: 'admin',
    access: 'canAdmin',
    routes: [
      {
        path: '/admin/sub-page',
        name: 'sub-page',
        exact: true,
        unaccessible: false,
        routes: [
          {
            path: '/admin/sub-page/list',
            name: 'sub-page-list',
            exact: true,
            unaccessible: false,
          },
        ],
      },
    ],
  },
  {
    name: 'list.table.result',
    path: '/list/:id',
    exact: true,
    unaccessible: false,
  },
  {
    name: 'list.table-list',
    path: '/list',
    exact: true,
    unaccessible: false,
  },
  { path: '/', redirect: '/welcome', exact: true, unaccessible: false },
];

const { menuData } = transformRoute(routes, true, ({ id }) => {
  if (id === 'menu.list.table-list') return '查询表格';
  if (id === 'menu.list.table.result') return '数据详情';
  if (id === 'menu.admin') return '管理页';
  if (id === 'menu.admin.sub-page') return '二级管理页';
  if (id === 'menu.admin.sub-page.sub-page-list') return '三级管理页';
  if (id === 'menu.welcome') return '欢迎';
  return id;
});

test('normal', () => {
  const openMenuItems = getMatchMenu('/admin/sub-page', menuData);
  expect(openMenuItems.length).toEqual(2);

  expect(openMenuItems[0].name).toEqual('管理页');
  expect(openMenuItems[1].name).toEqual('二级管理页');
  expect(openMenuItems).toMatchSnapshot();
});

test('three path', () => {
  const openMenuItems = getMatchMenu('/admin/sub-page/list', menuData);
  expect(openMenuItems.length).toEqual(3);

  expect(openMenuItems[0].name).toEqual('管理页');
  expect(openMenuItems[1].name).toEqual('二级管理页');
  expect(openMenuItems[2].name).toEqual('三级管理页');
  expect(openMenuItems).toMatchSnapshot();
});

test('var path', () => {
  const openMenuItems = getMatchMenu('/list/1234', menuData);
  expect(openMenuItems.length).toEqual(1);
  expect(openMenuItems[0].name).toEqual('数据详情');
  expect(openMenuItems).toMatchSnapshot();
});

test('test router', () => {
  const openMenuItems = getMatchMenu('/welcome/repertoryFw', testMenuData);
  expect(openMenuItems).toMatchSnapshot();
});

test('user path test', () => {
  const { menuData: userMenuData } = transformRoute(
    [
      {
        name: '人员组织管理',
        path: '/admin/userMng/',
        children: [
          { name: '人员管理', path: '/admin/userMng/users' },
          { name: '单位管理', path: '/admin/userMng/companies' },
        ],
      },
    ],
    false,
  );
  const openMenuItems = getMatchMenu('/admin/userMng/companies', userMenuData);
  expect(openMenuItems.length).toEqual(2);
  expect(openMenuItems).toMatchSnapshot();
});

test('test_config_router2', () => {
  const { menuData: userMenuData } = transformRoute(test_config_router2, false);
  const openMenuItems = getMatchMenu('/experienceScenario', userMenuData);
  expect(openMenuItems.length).toEqual(1);
  expect(openMenuItems).toMatchSnapshot();
});
