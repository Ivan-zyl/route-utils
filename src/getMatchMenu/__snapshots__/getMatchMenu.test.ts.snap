// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`normal 1`] = `
Array [
  Object {
    "access": "canAdmin",
    "children": Array [
      Object {
        "access": "canAdmin",
        "children": undefined,
        "exact": true,
        "key": "/admin/sub-page",
        "locale": "menu.admin.sub-page",
        "name": "二级管理页",
        "parentKeys": Array [
          "/",
          "/admin",
        ],
        "path": "/admin/sub-page",
        "routes": null,
        "unaccessible": false,
      },
    ],
    "key": "/admin",
    "locale": "menu.admin",
    "name": "管理页",
    "parentKeys": Array [
      "/",
    ],
    "path": "/admin",
    "routes": null,
  },
  Object {
    "access": "canAdmin",
    "children": undefined,
    "exact": true,
    "key": "/admin/sub-page",
    "locale": "menu.admin.sub-page",
    "name": "二级管理页",
    "parentKeys": Array [
      "/",
      "/admin",
    ],
    "path": "/admin/sub-page",
    "routes": null,
    "unaccessible": false,
  },
]
`;

exports[`var path 1`] = `
Array [
  Object {
    "children": undefined,
    "exact": true,
    "key": "/list/:id",
    "locale": "menu.list.table.result",
    "name": "数据详情",
    "parentKeys": Array [
      "/",
    ],
    "path": "/list/:id",
    "routes": null,
    "unaccessible": false,
  },
]
`;
